kong/runloop/handler.lua

```diff
@@ -152,95 +767,233 @@ do
+      end
     end
 
     return true
   end
 
+  local function rebuild_sync(callback, version)
+    kong.db:connect()
 
-  -- for unit-testing purposes only
-  _set_check_router_rebuild = function(f)
-    check_router_rebuild = f
+    local pok, ok, err = pcall(callback, version)
+    if not pok or not ok then
+      log(CRIT, "could not rebuild synchronously: ", ok or err)
+    end
+
+    kong.db:setkeepalive()
   end
 
+  local function rebuild_timer(premature, callback, version, semaphore)
+    if premature then
+      release_semaphore(semaphore)
+      return
+    end
 
-  get_router = function()
-    local version, err = singletons.cache:get("router:version",
-                                              CACHE_ROUTER_OPTS,
-                                              utils.uuid)
-    if err then
-      log(ngx.CRIT, "could not ensure router is up to date: ", err)
-      return nil, err
+    kong.db:connect()
+
+    local pok, ok, err = pcall(callback, version, true)
+    if not pok or not ok then
+      log(CRIT, "could not rebuild asynchronously: ", ok or err)
     end
 
-    if version == router_version then
-      return router
+    release_semaphore(semaphore)
+
+    kong.db:setkeepalive()
+  end
+
+  local function rebuild_async(callback, version, semaphore)
+    local ok, err = ngx.timer.at(0, rebuild_timer, callback, version, semaphore)
+    if not ok then
+      log(CRIT, "could not create rebuild timer: ", err)
+      return false
     end
 
-    -- wrap router rebuilds in a per-worker mutex:
-    -- this prevents dogpiling the database during rebuilds in
-    -- high-concurrency traffic patterns;
-    -- requests that arrive on this process during a router rebuild will be
-    -- queued. once the lock is acquired we re-check the
-    -- router version again to prevent unnecessary subsequent rebuilds
+    return true
+  end
 
-    local opts = {
-      name = "build_router",
-      timeout = build_router_timeout,
-      on_timeout = "run_unlocked",
-    }
-    local ok, err = concurrency.with_coroutine_mutex(opts, check_router_rebuild)
+  local function rebuild(name, callback, version, semaphore, wait)
+    local current_version = get_version(name)
+    if current_version == version then
+      return
+    end
 
+    local ok = acquire_semaphore(semaphore, wait)
     if not ok then
-      return nil, err
+      if wait and wait > 0 then
+        rebuild_sync(callback, current_version, semaphore)
+      end
+
+      return
+    end
+
+    current_version = get_version(name)
+    if current_version == version then
+      release_semaphore(semaphore)
+      return
+    end
+
+    if wait and wait > 0 then
+      rebuild_sync(callback, current_version, semaphore)
+      release_semaphore(semaphore)
+
+    else
+      ok = rebuild_async(callback, current_version, semaphore)
+      if not ok and wait == 0 then
+        rebuild_sync(callback, current_version, semaphore)
+        release_semaphore(semaphore)
+      end
```

Comment:
are we missing a `release_semaphore` in this branch?

Reply:
@p0pr0ck5 I think we are. Good catch.

Reply:
Fixed now.